File: app/lib/core/epub/epub_parser.dart
import 'dart:typed_data';

import 'package:archive/archive.dart';
import 'package:xml/xml.dart';

import 'epub_document.dart';

/// Parses EPUB files (both EPUB 2 and EPUB 3) from raw ZIP bytes.
class EpubParser {
  /// Parse an EPUB from raw bytes.
  static EpubDocument parse(Uint8List bytes) {
    final archive = ZipDecoder().decodeBytes(bytes);

    // 1. Locate the OPF file via META-INF/container.xml
    final opfPath = _findOpfPath(archive);
    final opfDir = _dirname(opfPath);

    // 2. Read and parse the OPF
    final opfXml = _readTextFile(archive, opfPath);
    final opfDoc = XmlDocument.parse(opfXml);

    // 3. Parse manifest
    final manifest = _parseManifest(opfDoc);
    final manifestById = {for (final m in manifest) m.id: m};

    // 4. Parse metadata
    final metadata = _parseMetadata(opfDoc, manifestById);

    // 5. Parse spine (resolving against manifest)
    final spine = _parseSpine(opfDoc, manifestById);

    // 6. Parse TOC (try EPUB 3 nav.xhtml first, fall back to NCX)
    final toc = _parseToc(archive, manifest, opfDir, opfDoc);

    // 7. Load all resources from the ZIP
    final resources = _loadResources(archive);

    return EpubDocument(
      metadata: metadata,
      manifest: manifest,
      spine: spine,
      tableOfContents: toc,
      resources: resources,
      opfDirectory: opfDir,
    );
  }

  // -----------------------------------------------------------------------
  // Container
  // -----------------------------------------------------------------------

  static String _findOpfPath(Archive archive) {
    final containerFile = archive.files.firstWhere(
      (f) => f.name == 'META-INF/container.xml',
      orElse: () => throw FormatException('No META-INF/container.xml found'),
    );
    final xml = String.fromCharCodes(containerFile.content as List<int>);
    final doc = XmlDocument.parse(xml);
    final rootfile = doc.findAllElements('rootfile').firstOrNull;
    final path = rootfile?.getAttribute('full-path');
    if (path == null || path.isEmpty) {
      throw const FormatException('No rootfile full-path in container.xml');
    }
    return path;
  }

  // -----------------------------------------------------------------------
  // Metadata
  // -----------------------------------------------------------------------

  static EpubMetadata _parseMetadata(
    XmlDocument opfDoc,
    Map<String, ManifestItem> manifestById,
  ) {
    String title = 'Unknown Title';
    final authors = <String>[];
    String? description;
    String? publisher;
    String? language;
    String? isbn;
    String? coverManifestId;

    // Find the <metadata> element (may be namespaced)
    final metadataEl = opfDoc.findAllElements('metadata').firstOrNull;
    if (metadataEl != null) {
      // Title
      final titleEl =
          metadataEl.findAllElements('dc:title').firstOrNull ??
          metadataEl.findAllElements('title').firstOrNull;
      if (titleEl != null && titleEl.innerText.isNotEmpty) {
        title = titleEl.innerText.trim();
      }

      // Authors (dc:creator)
      for (final el in [
        ...metadataEl.findAllElements('dc:creator'),
        ...metadataEl.findAllElements('creator'),
      ]) {
        final name = el.innerText.trim();
        if (name.isNotEmpty && !authors.contains(name)) authors.add(name);
      }

      // Description
      final descEl =
          metadataEl.findAllElements('dc:description').firstOrNull ??
          metadataEl.findAllElements('description').firstOrNull;
      description = descEl?.innerText.trim();
      if (description?.isEmpty == true) description = null;

      // Publisher
      final pubEl =
          metadataEl.findAllElements('dc:publisher').firstOrNull ??
          metadataEl.findAllElements('publisher').firstOrNull;
      publisher = pubEl?.innerText.trim();
      if (publisher?.isEmpty == true) publisher = null;

      // Language
      final langEl =
          metadataEl.findAllElements('dc:language').firstOrNull ??
          metadataEl.findAllElements('language').firstOrNull;
      language = langEl?.innerText.trim();
      if (language?.isEmpty == true) language = null;

      // ISBN from dc:identifier
      for (final el in [
        ...metadataEl.findAllElements('dc:identifier'),
        ...metadataEl.findAllElements('identifier'),
      ]) {
        final scheme =
            el.getAttribute('opf:scheme')?.toLowerCase() ??
            el.getAttribute('scheme')?.toLowerCase() ??
            '';
        final idAttr = el.getAttribute('id')?.toLowerCase() ?? '';
        if (scheme == 'isbn' || idAttr.contains('isbn')) {
          isbn = el.innerText.trim();
          break;
        }
      }

      // Cover image: <meta name="cover" content="manifest-id"> (EPUB 2)
      for (final meta in metadataEl.findAllElements('meta')) {
        if (meta.getAttribute('name')?.toLowerCase() == 'cover') {
          coverManifestId = meta.getAttribute('content');
          break;
        }
      }
    }

    // EPUB 3 cover: manifest item with properties="cover-image"
    if (coverManifestId == null) {
      for (final item in manifestById.values) {
        if (item.isCoverImage) {
          coverManifestId = item.id;
          break;
        }
      }
    }

    return EpubMetadata(
      title: title,
      authors: authors.isEmpty ? ['Unknown Author'] : authors,
      description: description,
      publisher: publisher,
      language: language,
      isbn: isbn,
      coverManifestId: coverManifestId,
    );
  }

  // -----------------------------------------------------------------------
  // Manifest
  // -----------------------------------------------------------------------

  static List<ManifestItem> _parseManifest(XmlDocument opfDoc) {
    final items = <ManifestItem>[];
    for (final el in opfDoc.findAllElements('item')) {
      final id = el.getAttribute('id');
      final href = el.getAttribute('href');
      final mediaType = el.getAttribute('media-type');
      if (id == null || href == null || mediaType == null) continue;

      items.add(ManifestItem(
        id: id,
        href: Uri.decodeFull(href),
        mediaType: mediaType,
        properties: el.getAttribute('properties'),
      ));
    }
    return items;
  }

  // -----------------------------------------------------------------------
  // Spine
  // -----------------------------------------------------------------------

  static List<SpineEntry> _parseSpine(
    XmlDocument opfDoc,
    Map<String, ManifestItem> manifestById,
  ) {
    final spineEl = opfDoc.findAllElements('spine').firstOrNull;
    if (spineEl == null) return [];

    final entries = <SpineEntry>[];
    for (final ref in spineEl.findElements('itemref')) {
      final idRef = ref.getAttribute('idref');
      if (idRef == null) continue;

      final manifestItem = manifestById[idRef];
      if (manifestItem == null) continue;

      // Only include HTML/XHTML content
      if (!manifestItem.isHtml) continue;

      final linearAttr = ref.getAttribute('linear');
      // Per spec: linear defaults to "yes" if absent
      final isLinear =
          linearAttr == null || linearAttr.toLowerCase() != 'no';

      entries.add(SpineEntry(
        idRef: idRef,
        isLinear: isLinear,
        href: manifestItem.href,
        mediaType: manifestItem.mediaType,
      ));
    }
    return entries;
  }

  // -----------------------------------------------------------------------
  // TOC — tries EPUB 3 nav.xhtml first, then EPUB 2 NCX
  // -----------------------------------------------------------------------

  static List<TocEntry> _parseToc(
    Archive archive,
    List<ManifestItem> manifest,
    String opfDir,
    XmlDocument opfDoc,
  ) {
    // EPUB 3: look for manifest item with properties="nav"
    final navItem = manifest.where((m) => m.isNav).firstOrNull;
    if (navItem != null) {
      final navPath = _resolveHref(navItem.href, opfDir);
      final navFile = _findFile(archive, navPath);
      if (navFile != null) {
        final navXml = String.fromCharCodes(navFile.content as List<int>);
        final entries = _parseNavXhtml(navXml);
        if (entries.isNotEmpty) return entries;
      }
    }

    // EPUB 2: find .ncx file (referenced by spine toc attribute or by media-type)
    final spineEl = opfDoc.findAllElements('spine').firstOrNull;
    final tocId = spineEl?.getAttribute('toc');
    ManifestItem? ncxItem;
    if (tocId != null) {
      ncxItem = manifest.where((m) => m.id == tocId).firstOrNull;
    }
    ncxItem ??= manifest
        .where((m) => m.mediaType == 'application/x-dtbncx+xml')
        .firstOrNull;

    if (ncxItem != null) {
      final ncxPath = _resolveHref(ncxItem.href, opfDir);
      final ncxFile = _findFile(archive, ncxPath);
      if (ncxFile != null) {
        final ncxXml = String.fromCharCodes(ncxFile.content as List<int>);
        return _parseNcx(ncxXml);
      }
    }

    return [];
  }

  /// Parse an EPUB 3 navigation document (nav.xhtml).
  static List<TocEntry> _parseNavXhtml(String xhtml) {
    final doc = XmlDocument.parse(xhtml);
    // Find the <nav epub:type="toc"> element
    for (final nav in doc.findAllElements('nav')) {
      final epubType =
          nav.getAttribute('epub:type') ?? nav.getAttribute('type') ?? '';
      if (epubType.contains('toc')) {
        final ol = nav.findElements('ol').firstOrNull;
        if (ol != null) return _parseNavOl(ol);
      }
    }
    return [];
  }

  static List<TocEntry> _parseNavOl(XmlElement ol) {
    final entries = <TocEntry>[];
    for (final li in ol.findElements('li')) {
      final a = li.findElements('a').firstOrNull;
      if (a == null) continue;

      final fullHref = a.getAttribute('href') ?? '';
      final parts = fullHref.split('#');
      final href = Uri.decodeFull(parts[0]);
      final fragment = parts.length > 1 ? parts[1] : null;
      final title = a.innerText.trim();

      // Check for nested <ol>
      final childOl = li.findElements('ol').firstOrNull;
      final children = childOl != null ? _parseNavOl(childOl) : <TocEntry>[];

      if (title.isNotEmpty) {
        entries.add(TocEntry(
          title: title,
          href: href,
          fragment: fragment,
          fullHref: fullHref,
          children: children,
        ));
      }
    }
    return entries;
  }

  /// Parse an EPUB 2 NCX TOC.
  static List<TocEntry> _parseNcx(String ncxXml) {
    final doc = XmlDocument.parse(ncxXml);
    final navMap = doc.findAllElements('navMap').firstOrNull;
    if (navMap == null) return [];

    return _parseNavPoints(navMap);
  }

  static List<TocEntry> _parseNavPoints(XmlElement parent) {
    final entries = <TocEntry>[];
    for (final np in parent.findElements('navPoint')) {
      final label = np
          .findElements('navLabel')
          .firstOrNull
          ?.findElements('text')
          .firstOrNull
          ?.innerText
          .trim();
      final src =
          np.findElements('content').firstOrNull?.getAttribute('src') ?? '';

      if (label == null || label.isEmpty) continue;

      final decodedSrc = Uri.decodeFull(src);
      final parts = decodedSrc.split('#');
      final href = parts[0];
      final fragment = parts.length > 1 ? parts[1] : null;

      final children = _parseNavPoints(np);

      entries.add(TocEntry(
        title: label,
        href: href,
        fragment: fragment,
        fullHref: decodedSrc,
        children: children,
      ));
    }
    return entries;
  }

  // -----------------------------------------------------------------------
  // Resources
  // -----------------------------------------------------------------------

  static Map<String, EpubResource> _loadResources(Archive archive) {
    final resources = <String, EpubResource>{};
    for (final file in archive.files) {
      if (!file.isFile) continue;
      // Skip META-INF and mimetype
      if (file.name.startsWith('META-INF/') || file.name == 'mimetype') {
        continue;
      }

      final mediaType = _guessMediaType(file.name);
      resources[file.name] = EpubResource(
        href: file.name,
        mediaType: mediaType,
        data: Uint8List.fromList(file.content as List<int>),
      );
    }
    return resources;
  }

  // -----------------------------------------------------------------------
  // Helpers
  // -----------------------------------------------------------------------

  static String _readTextFile(Archive archive, String path) {
    final file = archive.files.firstWhere(
      (f) => f.name == path,
      orElse: () => throw FormatException('File not found in EPUB: $path'),
    );
    return String.fromCharCodes(file.content as List<int>);
  }

  static ArchiveFile? _findFile(Archive archive, String path) {
    for (final f in archive.files) {
      if (f.name == path) return f;
    }
    // Try case-insensitive and with/without leading slash
    final normalized = path.startsWith('/') ? path.substring(1) : path;
    for (final f in archive.files) {
      if (f.name.toLowerCase() == normalized.toLowerCase()) return f;
    }
    return null;
  }

  static String _dirname(String path) {
    final lastSlash = path.lastIndexOf('/');
    return lastSlash >= 0 ? path.substring(0, lastSlash + 1) : '';
  }

  static String _resolveHref(String href, String baseDir) {
    if (href.startsWith('/') || href.startsWith('http')) return href;
    return '$baseDir$href';
  }

  static String _guessMediaType(String path) {
    final lower = path.toLowerCase();
    if (lower.endsWith('.xhtml') || lower.endsWith('.html') || lower.endsWith('.htm')) {
      return 'application/xhtml+xml';
    }
    if (lower.endsWith('.css')) return 'text/css';
    if (lower.endsWith('.jpg') || lower.endsWith('.jpeg')) return 'image/jpeg';
    if (lower.endsWith('.png')) return 'image/png';
    if (lower.endsWith('.gif')) return 'image/gif';
    if (lower.endsWith('.svg')) return 'image/svg+xml';
    if (lower.endsWith('.webp')) return 'image/webp';
    if (lower.endsWith('.ttf')) return 'font/ttf';
    if (lower.endsWith('.otf')) return 'font/otf';
    if (lower.endsWith('.woff')) return 'font/woff';
    if (lower.endsWith('.woff2')) return 'font/woff2';
    if (lower.endsWith('.ncx')) return 'application/x-dtbncx+xml';
    if (lower.endsWith('.opf')) return 'application/oebps-package+xml';
    if (lower.endsWith('.xml')) return 'application/xml';
    if (lower.endsWith('.js')) return 'application/javascript';
    return 'application/octet-stream';
  }
}

File: macos-app/Moku/EPUB/EpubParser.swift
import Foundation
import ZIPFoundation

enum EpubParserError: LocalizedError {
    case invalidArchive
    case containerNotFound
    case opfPathNotFound
    case opfNotFound
    case parseError(String)

    var errorDescription: String? {
        switch self {
        case .invalidArchive: "The file is not a valid EPUB archive."
        case .containerNotFound: "META-INF/container.xml not found."
        case .opfPathNotFound: "Could not locate OPF path in container.xml."
        case .opfNotFound: "OPF file not found in archive."
        case .parseError(let msg): "EPUB parse error: \(msg)"
        }
    }
}

enum EpubParser {

    // MARK: - Public

    static func parse(data: Data) throws -> EpubDocument {
        guard let archive = Archive(data: data, accessMode: .read) else {
            throw EpubParserError.invalidArchive
        }

        let opfPath = try findOpfPath(archive: archive)
        let opfDirectory = (opfPath as NSString).deletingLastPathComponent
        let opfData = try extractEntry(archive: archive, path: opfPath)

        let opfDoc = try XMLDocument(data: opfData, options: [.nodePreserveAll])

        let manifest = parseManifest(opfDoc: opfDoc)
        let manifestById = Dictionary(manifest.map { ($0.id, $0) }, uniquingKeysWith: { a, _ in a })

        let spine = parseSpine(opfDoc: opfDoc, manifestById: manifestById)
        let metadata = parseMetadata(opfDoc: opfDoc, manifestById: manifestById)
        let toc = parseToc(
            archive: archive,
            manifest: manifest,
            opfDirectory: opfDirectory,
            opfDoc: opfDoc
        )
        let resources = loadResources(archive: archive)

        return EpubDocument(
            metadata: metadata,
            manifest: manifest,
            spine: spine,
            toc: toc,
            resources: resources,
            opfDirectory: opfDirectory
        )
    }

    // MARK: - OPF Location

    static func findOpfPath(archive: Archive) throws -> String {
        let containerData = try extractEntry(archive: archive, path: "META-INF/container.xml")
        let containerDoc = try XMLDocument(data: containerData, options: [.nodePreserveAll])

        // Try with namespace first, then without
        let xpaths = [
            "//container:rootfile/@full-path",
            "//rootfile/@full-path",
        ]
        let namespaces = ["container": "urn:oasis:names:tc:opendocument:xmlns:container"]

        for xpath in xpaths {
            if let nodes = try? containerDoc.nodes(forXPath: xpath) {
                for node in nodes {
                    if let path = node.stringValue, !path.isEmpty {
                        return path
                    }
                }
            }
            // Try with explicit namespace mapping
            if xpath.contains("container:") {
                let nsXpath = xpath
                if let root = containerDoc.rootElement() {
                    for (prefix, uri) in namespaces {
                        root.addNamespace(XMLNode.namespace(withName: prefix, stringValue: uri) as! XMLNode)
                    }
                }
                if let nodes = try? containerDoc.nodes(forXPath: nsXpath) {
                    for node in nodes {
                        if let path = node.stringValue, !path.isEmpty {
                            return path
                        }
                    }
                }
            }
        }

        // Fallback: scan for .opf files in archive
        for entry in archive {
            if entry.path.hasSuffix(".opf") {
                return entry.path
            }
        }

        throw EpubParserError.opfPathNotFound
    }

    // MARK: - Metadata

    static func parseMetadata(opfDoc: XMLDocument, manifestById: [String: ManifestItem]) -> EpubMetadata {
        let root = opfDoc.rootElement()

        // Helper to find text in common metadata XPaths
        func metaText(_ localName: String) -> String? {
            let xpaths = [
                "//dc:\(localName)",
                "//opf:metadata/dc:\(localName)",
                "//*[local-name()='\(localName)']",
            ]
            for xpath in xpaths {
                if let nodes = try? opfDoc.nodes(forXPath: xpath) {
                    for node in nodes {
                        if let text = node.stringValue, !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                            return text.trimmingCharacters(in: .whitespacesAndNewlines)
                        }
                    }
                }
            }
            return nil
        }

        func allMetaText(_ localName: String) -> [String] {
            let xpath = "//*[local-name()='\(localName)']"
            guard let nodes = try? opfDoc.nodes(forXPath: xpath) else { return [] }
            return nodes.compactMap { node in
                guard let text = node.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines),
                      !text.isEmpty else { return nil }
                return text
            }
        }

        let title = metaText("title") ?? "Untitled"
        var authors = allMetaText("creator")
        if authors.isEmpty, let single = metaText("creator") {
            authors = [single]
        }

        let description = metaText("description")
        let publisher = metaText("publisher")
        let language = metaText("language")

        // ISBN from dc:identifier
        var isbn: String?
        let identifiers = allMetaText("identifier")
        for id in identifiers {
            let cleaned = id.replacingOccurrences(of: "-", with: "")
            if cleaned.range(of: #"^\d{10}(\d{3})?$"#, options: .regularExpression) != nil {
                isbn = id
                break
            }
        }

        // Cover manifest ID from <meta name="cover" content="...">
        var coverManifestId: String?
        if let metaNodes = try? opfDoc.nodes(forXPath: "//*[local-name()='meta']") {
            for node in metaNodes {
                if let element = node as? XMLElement,
                   element.attribute(forName: "name")?.stringValue == "cover",
                   let content = element.attribute(forName: "content")?.stringValue
                {
                    coverManifestId = content
                    break
                }
            }
        }

        // Also check for cover-image property in manifest
        if coverManifestId == nil {
            for (id, item) in manifestById where item.isCoverImage {
                coverManifestId = id
                break
            }
        }

        return EpubMetadata(
            title: title,
            authors: authors,
            description: description,
            publisher: publisher,
            language: language,
            isbn: isbn,
            coverManifestId: coverManifestId
        )
    }

    // MARK: - Manifest

    static func parseManifest(opfDoc: XMLDocument) -> [ManifestItem] {
        var items: [ManifestItem] = []
        let xpaths = [
            "//*[local-name()='manifest']/*[local-name()='item']",
            "//manifest/item",
        ]

        for xpath in xpaths {
            guard let nodes = try? opfDoc.nodes(forXPath: xpath) else { continue }
            for node in nodes {
                guard let element = node as? XMLElement,
                      let id = element.attribute(forName: "id")?.stringValue,
                      let href = element.attribute(forName: "href")?.stringValue,
                      let mediaType = element.attribute(forName: "media-type")?.stringValue
                else { continue }

                let properties = element.attribute(forName: "properties")?.stringValue
                items.append(ManifestItem(id: id, href: href, mediaType: mediaType, properties: properties))
            }
            if !items.isEmpty { break }
        }

        return items
    }

    // MARK: - Spine

    static func parseSpine(opfDoc: XMLDocument, manifestById: [String: ManifestItem]) -> [SpineEntry] {
        var entries: [SpineEntry] = []
        let xpaths = [
            "//*[local-name()='spine']/*[local-name()='itemref']",
            "//spine/itemref",
        ]

        for xpath in xpaths {
            guard let nodes = try? opfDoc.nodes(forXPath: xpath) else { continue }
            for node in nodes {
                guard let element = node as? XMLElement,
                      let idRef = element.attribute(forName: "idref")?.stringValue
                else { continue }

                let linear = element.attribute(forName: "linear")?.stringValue != "no"
                let item = manifestById[idRef]

                entries.append(SpineEntry(
                    idRef: idRef,
                    isLinear: linear,
                    href: item?.href ?? "",
                    mediaType: item?.mediaType ?? "application/xhtml+xml"
                ))
            }
            if !entries.isEmpty { break }
        }

        return entries
    }

    // MARK: - TOC

    static func parseToc(
        archive: Archive,
        manifest: [ManifestItem],
        opfDirectory: String,
        opfDoc: XMLDocument
    ) -> [TocEntry] {
        // EPUB 3: try nav document first
        if let navItem = manifest.first(where: { $0.isNav }) {
            let navPath = opfDirectory.isEmpty ? navItem.href : opfDirectory + "/" + navItem.href
            if let navData = try? extractEntry(archive: archive, path: navPath),
               let navString = String(data: navData, encoding: .utf8)
            {
                let entries = parseNavXhtml(xhtml: navString, basePath: (navPath as NSString).deletingLastPathComponent)
                if !entries.isEmpty { return entries }
            }
        }

        // EPUB 2 fallback: NCX
        // Find NCX from spine toc attribute or manifest
        var ncxItem: ManifestItem?

        // Check spine element for toc attribute
        if let spineNodes = try? opfDoc.nodes(forXPath: "//*[local-name()='spine']"),
           let spineElement = spineNodes.first as? XMLElement,
           let tocId = spineElement.attribute(forName: "toc")?.stringValue
        {
            ncxItem = manifest.first { $0.id == tocId }
        }

        // Fallback: look for NCX media type
        if ncxItem == nil {
            ncxItem = manifest.first { $0.mediaType == "application/x-dtbncx+xml" }
        }

        if let ncxItem = ncxItem {
            let ncxPath = opfDirectory.isEmpty ? ncxItem.href : opfDirectory + "/" + ncxItem.href
            if let ncxData = try? extractEntry(archive: archive, path: ncxPath),
               let ncxString = String(data: ncxData, encoding: .utf8)
            {
                return parseNcx(ncxXml: ncxString, basePath: (ncxPath as NSString).deletingLastPathComponent)
            }
        }

        return []
    }

    /// Parse EPUB 3 nav XHTML into TocEntry tree.
    static func parseNavXhtml(xhtml: String, basePath: String = "") -> [TocEntry] {
        guard let data = xhtml.data(using: .utf8),
              let doc = try? XMLDocument(data: data, options: [.nodePreserveAll, .documentTidyHTML])
        else { return [] }

        // Find the <nav epub:type="toc"> or first <nav>
        let navXpaths = [
            "//*[local-name()='nav'][@*[local-name()='type']='toc']",
            "//*[local-name()='nav']",
        ]

        var navElement: XMLElement?
        for xpath in navXpaths {
            if let nodes = try? doc.nodes(forXPath: xpath), let el = nodes.first as? XMLElement {
                navElement = el
                break
            }
        }

        guard let nav = navElement else { return [] }

        // Find the top-level <ol> inside nav
        guard let olNodes = try? nav.nodes(forXPath: "*[local-name()='ol']"),
              let ol = olNodes.first as? XMLElement
        else { return [] }

        return parseNavOl(ol: ol, basePath: basePath)
    }

    private static func parseNavOl(ol: XMLElement, basePath: String) -> [TocEntry] {
        var entries: [TocEntry] = []

        guard let liNodes = try? ol.nodes(forXPath: "*[local-name()='li']") else {
            return entries
        }

        for liNode in liNodes {
            guard let li = liNode as? XMLElement else { continue }

            // Get <a> element
            guard let aNodes = try? li.nodes(forXPath: "*[local-name()='a']"),
                  let a = aNodes.first as? XMLElement,
                  let rawHref = a.attribute(forName: "href")?.stringValue
            else { continue }

            let title = a.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !title.isEmpty else { continue }

            let fullHref = resolveHref(rawHref, basePath: basePath)
            let (path, fragment) = splitFragment(fullHref)

            // Recurse into nested <ol>
            var children: [TocEntry] = []
            if let subOlNodes = try? li.nodes(forXPath: "*[local-name()='ol']"),
               let subOl = subOlNodes.first as? XMLElement
            {
                children = parseNavOl(ol: subOl, basePath: basePath)
            }

            entries.append(TocEntry(
                title: title,
                href: path,
                fragment: fragment,
                fullHref: fullHref,
                children: children
            ))
        }

        return entries
    }

    /// Parse EPUB 2 NCX into TocEntry tree.
    static func parseNcx(ncxXml: String, basePath: String = "") -> [TocEntry] {
        guard let data = ncxXml.data(using: .utf8),
              let doc = try? XMLDocument(data: data, options: [.nodePreserveAll])
        else { return [] }

        // Find <navMap>
        guard let navMapNodes = try? doc.nodes(forXPath: "//*[local-name()='navMap']"),
              let navMap = navMapNodes.first as? XMLElement
        else { return [] }

        return parseNcxNavPoints(parent: navMap, basePath: basePath)
    }

    private static func parseNcxNavPoints(parent: XMLElement, basePath: String) -> [TocEntry] {
        var entries: [TocEntry] = []

        guard let navPoints = try? parent.nodes(forXPath: "*[local-name()='navPoint']") else {
            return entries
        }

        for npNode in navPoints {
            guard let np = npNode as? XMLElement else { continue }

            // Get title from <navLabel><text>
            let title: String
            if let textNodes = try? np.nodes(forXPath: "*[local-name()='navLabel']/*[local-name()='text']"),
               let text = textNodes.first?.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines),
               !text.isEmpty
            {
                title = text
            } else {
                continue
            }

            // Get href from <content src="...">
            guard let contentNodes = try? np.nodes(forXPath: "*[local-name()='content']"),
                  let contentEl = contentNodes.first as? XMLElement,
                  let rawSrc = contentEl.attribute(forName: "src")?.stringValue
            else { continue }

            let fullHref = resolveHref(rawSrc, basePath: basePath)
            let (path, fragment) = splitFragment(fullHref)

            let children = parseNcxNavPoints(parent: np, basePath: basePath)

            entries.append(TocEntry(
                title: title,
                href: path,
                fragment: fragment,
                fullHref: fullHref,
                children: children
            ))
        }

        return entries
    }

    // MARK: - Resources

    static func loadResources(archive: Archive) -> [String: EpubResource] {
        var resources: [String: EpubResource] = [:]

        for entry in archive where entry.type == .file {
            let path = entry.path
            guard !path.hasPrefix("META-INF/"), path != "mimetype" else { continue }

            var entryData = Data()
            do {
                _ = try archive.extract(entry) { chunk in entryData.append(chunk) }
            } catch {
                continue
            }

            let mediaType = guessMediaType(path: path)
            resources[path] = EpubResource(href: path, mediaType: mediaType, data: entryData)
        }

        return resources
    }

    // MARK: - Helpers

    static func guessMediaType(path: String) -> String {
        let ext = (path as NSString).pathExtension.lowercased()
        switch ext {
        case "xhtml", "xht": return "application/xhtml+xml"
        case "html", "htm": return "text/html"
        case "css": return "text/css"
        case "js": return "application/javascript"
        case "jpg", "jpeg": return "image/jpeg"
        case "png": return "image/png"
        case "gif": return "image/gif"
        case "svg": return "image/svg+xml"
        case "webp": return "image/webp"
        case "ttf": return "font/ttf"
        case "otf": return "font/otf"
        case "woff": return "font/woff"
        case "woff2": return "font/woff2"
        case "xml": return "application/xml"
        case "ncx": return "application/x-dtbncx+xml"
        case "opf": return "application/oebps-package+xml"
        case "mp3": return "audio/mpeg"
        case "mp4": return "video/mp4"
        case "smil": return "application/smil+xml"
        default: return "application/octet-stream"
        }
    }

    private static func extractEntry(archive: Archive, path: String) throws -> Data {
        // Try exact path first
        if let entry = archive[path] {
            var data = Data()
            _ = try archive.extract(entry) { chunk in data.append(chunk) }
            return data
        }

        // Case-insensitive fallback
        let lowered = path.lowercased()
        for entry in archive where entry.path.lowercased() == lowered {
            var data = Data()
            _ = try archive.extract(entry) { chunk in data.append(chunk) }
            return data
        }

        throw EpubParserError.containerNotFound
    }

    private static func resolveHref(_ href: String, basePath: String) -> String {
        let decoded = href.removingPercentEncoding ?? href
        if decoded.hasPrefix("/") || decoded.contains("://") {
            return decoded
        }
        if basePath.isEmpty {
            return decoded
        }
        // Resolve relative path
        let base = URL(fileURLWithPath: "/" + basePath, isDirectory: true)
        let resolved = URL(fileURLWithPath: decoded, relativeTo: base)
        var path = resolved.standardized.path
        if path.hasPrefix("/") {
            path = String(path.dropFirst())
        }
        return path
    }

    private static func splitFragment(_ href: String) -> (path: String, fragment: String?) {
        guard let hashIndex = href.firstIndex(of: "#") else {
            return (href, nil)
        }
        let path = String(href[href.startIndex..<hashIndex])
        let fragment = String(href[href.index(after: hashIndex)...])
        return (path, fragment.isEmpty ? nil : fragment)
    }
}
